dbms/src/Functions/FunctionsRegexp.h

```diff
@@ -0,0 +1,1078 @@
+#include <re2/re2.h>
+
+#include <memory>
+#include <type_traits>
+
+#include "Columns/ColumnNullable.h"
+#include "Columns/ColumnString.h"
+#include "Columns/ColumnsNumber.h"
+#include "Columns/IColumn.h"
+#include "Common/Exception.h"
+#include "Core/Field.h"
+#include "Core/Types.h"
+#include "DataTypes/DataTypeNothing.h"
+#include "DataTypes/DataTypeNullable.h"
+#include "Parsers/Lexer.h"
+#include "common/StringRef.h"
+#include "common/defines.h"
+#include "common/types.h"
+
+#if USE_RE2_ST
+#include <re2_st/re2.h>
+#else
+#define re2_st re2
+#endif
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int BAD_ARGUMENTS;
+extern const int ILLEGAL_COLUMN;
+extern const int TOO_LESS_ARGUMENTS_FOR_FUNCTION;
+extern const int TOO_MANY_ARGUMENTS_FOR_FUNCTION;
+} // namespace ErrorCodes
+
+const char * EMPTY_PAT_ERR_MSG = "Empty pattern is invalid";
+
+struct NameTiDBRegexp
+{
+    static constexpr auto name = "regexp";
+};
+struct NameRegexpLike
+{
+    static constexpr auto name = "regexp_like";
+};
+struct NameReplaceOne
+{
+    static constexpr auto name = "replaceOne";
+};
+struct NameReplaceAll
+{
+    static constexpr auto name = "replaceAll";
+};
+struct NameReplaceRegexpOne
+{
+    static constexpr auto name = "replaceRegexpOne";
+};
+struct NameReplaceRegexpAll
+{
+    static constexpr auto name = "replaceRegexpAll";
+};
+
+static constexpr std::string_view regexp_name(NameTiDBRegexp::name);
+static constexpr std::string_view regexp_like_name(NameRegexpLike::name);
+
+String getMatchType(const String & match_type, TiDB::TiDBCollatorPtr collator = nullptr);
+
+inline int getDefaultFlags()
+{
+    int flags = 0;
+    flags |= OptimizedRegularExpressionImpl<false>::RE_NO_CAPTURE | OptimizedRegularExpressionImpl<false>::RE_NO_OPTIMIZE;
+    return flags;
+}
+
+inline String addMatchTypeForPattern(const String & pattern, const String & match_type, TiDB::TiDBCollatorPtr collator)
+{
+    String flags = getMatchType(match_type, collator);
+    if (flags.empty())
+        return pattern;
+    return fmt::format("(?{}){}", flags, pattern);
+}
+
+inline Regexps::Pool::Pointer createRegexpWithMatchType(const String & pattern, const String & match_type, TiDB::TiDBCollatorPtr collator)
+{
+    String final_pattern = addMatchTypeForPattern(pattern, match_type, collator);
+    return Regexps::get<false, true>(final_pattern, getDefaultFlags());
+}
+
+// Only int types used in ColumnsNumber.h can be valid
+template <typename T>
+inline constexpr bool check_int_type()
+{
+    return static_cast<bool>(std::is_same_v<T, UInt8> || std::is_same_v<T, UInt16> || std::is_same_v<T, UInt32> || std::is_same_v<T, UInt64> || std::is_same_v<T, UInt128> || std::is_same_v<T, Int8> || std::is_same_v<T, Int16> || std::is_same_v<T, Int32> || std::is_same_v<T, Int64>);
+}
+
+// Use this type when param is not provided
+class ParamDefault
+{
+public:
+    explicit ParamDefault(Int64 val)
+        : default_int(val)
+        , default_string("")
+    {}
```

Comment:
what happens if `val` is not 0?

Reply:
> what happens if `val` is not 0?

All ParamDefault has been replaced by const ParamString or const ParamInt